jamaljsr/polar
# feat(images): add support for custom docker images

Closes #255, #288 

### Description

This PR adds the ability to create custom nodes which use a specified docker image other than the ones managed in the polarlightning Docker Hub repo. It also adds the ability to customize the docker startup command for both custom and managed nodes. You can customize the commands globally for all future nodes, or for a specific node in an existing network.

There is a new Manage Nodes screen where you can create, edit & delete your custom nodes. You can also alter the default command for all Polar managed nodes. When creating a new network, you can specify how many custom nodes you want to add. Once a network is created, you can also edit the startup command used by each node individually.

I still need to add a bunch more unit tests, but all of the functionality is there. I just want to get this feature out to get feedback from from those that requested it. Please let me know if this solves your use cases @arvchristos & @torkelrogstad.

### Steps to Test

Build a custom image for one of the supported implementations. For LND master branch, follow the steps in [this gist](https://gist.github.com/jamaljsr/f7fe765a0f7a51d4333adc570cf6b3cd). Docs for other implementations will be completed before the next release.

1. Click on Manage Nodes in the navbar menu
1. Click on the Add Custom Node button
1. Fill in the form choosing lnd:master for the Docker Image and click Save
1. Click on Create Network in the navbar menu
1. Enter 1 for the number of your custom node and click Create Network
1. Click on the first node and confirm the sidebar Info tab displays your custom docker image name
1. Start the network
1. Open a terminal for the first node
1. Enter `lncli getinfo` to confirm the version the latest LND commit hash

### Screenshots

![image](https://user-images.githubusercontent.com/1356600/74966637-6c6e3700-53e5-11ea-9b35-d1ecf1d4b813.png)

There are many more screenshots collapsed below. I hid them to avoid the pain of scrolling 
<details>
<summary>View more screenshots</summary>

Add a new custom node
![image](https://user-images.githubusercontent.com/1356600/74966590-51032c00-53e5-11ea-9575-92f934f7d777.png)

Edit a managed node's command
![image](https://user-images.githubusercontent.com/1356600/74966898-fd451280-53e5-11ea-8c3a-4abadbe40ae4.png)

Create a new Network with custom nodes
![image](https://user-images.githubusercontent.com/1356600/74966962-14840000-53e6-11ea-844f-21ca349b9772.png)

Custom nodes are displayed in the designer sidebar
![image](https://user-images.githubusercontent.com/1356600/74967351-cd4a3f00-53e6-11ea-91be-cffc24f2e275.png)

New Advanced Options button in the Actions tab
![image](https://user-images.githubusercontent.com/1356600/74970322-2ec0dc80-53ec-11ea-8ab3-776efac8a3df.png)

Edit the command of an existing node
![image](https://user-images.githubusercontent.com/1356600/74967079-4d23d980-53e6-11ea-9df6-03a5a774f729.png)

</details>
